### PR TITLE
MultiNiftiLabelsMasker object for handling 4D + 1D input

### PR DESCRIPTION
**Pull request by Hororohoruru**
One of the objects proposed in #1240 , to adress several issues (e.g., #2247 ).

I used the MultiNiftiMasker object as a rough template for this. This is the very first version and needs a lot of improvements. For now, I added a test to check it accepts 4D + 1D input, and it is handled by parallelizing the function transform_single_imgs of NiftiLabelsMasker.

Several issues need to be addressed, I would be grateful for suggestions on how to do so:
- Since each 4D image needs the labels_img to be resample to its affine, there needs to be a way
of storing the affines/label images.
- For the 'data' strategy, it can be done by parallelizing the single transform method from the NiftiLabelsMasker, but the labels_ attribute gets overwritten after each iteration. This could be fixed by wrapping the transform_single_imgs function in another function that recovers this attribute and allocates it in a list for each iteration of the parallel loop (any ideas?)
- For the other strategies, that are handled at fit time, I don't know what would be the best way to work around it for the 4D + 1D case

Other questions:
- Inverse transform (not implemented) would have a similar problem: If several labels_img are needed, it could be a bit problematic afterwards.
- What else should be tested for this class?

### CODE REVIEW

**Comment by kchawla-pi** [comment]
Pull from master to fix CircleCI failure.

File: nilearn/input_data/multi_nifti_labels_masker.py [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+        The number of CPUs to use to do the computation. -1 means
+        'all CPUs', -2 'all CPUs but one', and so on.
+
+    verbose: integer, optional
+        Indicate the level of verbosity. By default, nothing is printed
+
+    strategy: str
+        The name of a valid function to reduce the region with.
+        Must be one of: sum, mean, median, mininum, maximum, variance,
+        standard_deviation
```
Comment (kchawla-pi) [inline nilearn/input_data/multi_nifti_labels_masker.py]: Mention the defaults in the docstring.

File: nilearn/input_data/multi_nifti_labels_masker.py [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+
+        available_reduction_strategies = {'mean', 'median', 'sum',
+                                          'minimum', 'maximum',
+                                          'standard_deviation', 'variance'}
+
+        if strategy not in available_reduction_strategies:
+            raise ValueError(str.format(
+                "Invalid strategy '{}'. Valid strategies are {}.",
+                strategy,
+                available_reduction_strategies
```
Comment (kchawla-pi) [inline nilearn/input_data/multi_nifti_labels_masker.py]: Instead of `str.format(`, write this as 
```python
"Invalid strategy '{}'. Valid strategies are {}.".format(strategy, available_reduction_strategies)
```
Of course, breaking it down to fit.

File: nilearn/input_data/multi_nifti_labels_masker.py [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+            ))
+
+        self.strategy = strategy
+
+        if resampling_target not in ("labels", "data", None):
+            raise ValueError("invalid value for 'resampling_target' "
+                             "parameter: " + str(resampling_target))
```
Comment (kchawla-pi) [inline nilearn/input_data/multi_nifti_labels_masker.py]:
```python
("invalid value for 'resampling_target' "
                             "parameter: {}".format(resampling_target))
```

Comment (kchawla-pi) [reply nilearn/input_data/multi_nifti_labels_masker.py]: Make the message same style as above.

File: nilearn/input_data/multi_nifti_labels_masker.py [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+            List of signals for each label per subject.
+            shape: list of (number of scans, number of labels)
+        """
+        # We handle the resampling of labels separately because the affine of
+        # the labels image should not impact the extraction of the signal.
+
+        if not hasattr(self, 'mask_img_'):
+            raise ValueError('It seems that %s has not been fitted. '
+                             'You must call fit() before calling transform().'
+                             % self.__class__.__name__)
```
Comment (kchawla-pi) [inline nilearn/input_data/multi_nifti_labels_masker.py]:
```python
('It seems that {} has not been fitted. '
'You must call fit() before calling transform().'.format(
                             self.__class__.__name__)
)
```

Comment (kchawla-pi) [reply nilearn/input_data/multi_nifti_labels_masker.py]: And so on, everywhere.

File: nilearn/input_data/multi_nifti_labels_masker.py [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+
+        Returns
+        -------
+        data: {list of numpy arrays}
+            preprocessed images
+        """
+
+        self._check_fitted()
+        if not hasattr(imgs, '__iter__') \
+                or isinstance(imgs, _basestring):
```
Comment (kchawla-pi) [inline nilearn/input_data/multi_nifti_labels_masker.py]:
```python
if (not hasattr(imgs, '__iter__')
    or isinstance(imgs, _basestring)): 
```
Use of `\` for line continuation is no longer desired in Python.

File: nilearn/input_data/tests/test_multi_nifti_labels_masker.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
+import numpy as np
+
+import nibabel
+import pytest
+
+from nilearn.input_data.nifti_labels_masker import NiftiLabelsMasker
+from nilearn.input_data.multi_nifti_labels_masker import MultiNiftiLabelsMasker
+from nilearn._utils import testing, as_ndarray, data_gen
```
Comment (kchawla-pi) [inline nilearn/input_data/tests/test_multi_nifti_labels_masker.py]: sort the imported objects alphabetically.

File: nilearn/input_data/tests/test_multi_nifti_labels_masker.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+from nilearn._utils.exceptions import DimensionError
+from nilearn.image import get_data
```
Comment (kchawla-pi) [inline nilearn/input_data/tests/test_multi_nifti_labels_masker.py]: Sort the imports alphabetically. _utils should be on top.

File: nilearn/input_data/tests/test_multi_nifti_labels_masker.py [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+
+
+def generate_random_img(shape, length=1, affine=np.eye(4),
```
Comment (kchawla-pi) [inline nilearn/input_data/tests/test_multi_nifti_labels_masker.py]: There are already functions in Nilearn to generate fake data. Use them instead of writing a new one.

File: nilearn/input_data/tests/test_multi_nifti_labels_masker.py [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+    signals11 = masker11.fit_transform(fmri11_img)
+    assert signals11.shape == (length, n_regions)
+
+    with pytest.raises(ValueError, match='has not been fitted. '):
+        NiftiLabelsMasker(labels11_img).inverse_transform(signals11)
+
+    # Call inverse transform (smoke test)
+    fmri11_img_r = masker11.inverse_transform(signals11)
+    assert fmri11_img_r.shape == fmri11_img.shape
+    np.testing.assert_almost_equal(fmri11_img_r.affine, fmri11_img.affine)
```
Comment (kchawla-pi) [inline nilearn/input_data/tests/test_multi_nifti_labels_masker.py]: There should be a new line at the end of file.

**Comment by kchawla-pi** [comment]
Made a quick pass initial review for small stuff.
haven't looked at functionality or details in it.
The code quality itself is pretty good.

File: nilearn/input_data/multi_nifti_labels_masker.py [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+from .._utils.niimg_conversions import _iter_check_niimg
+from .nifti_labels_masker import NiftiLabelsMasker
+from nilearn.image import get_data
+
+
+class MultiNiftiLabelsMasker(NiftiLabelsMasker, CacheMixin):
+    """Class for masking of Niimg-like objects.
+
+    MultiNiftiLabelsMasker is useful when data from non-overlapping volumes and from
+    different subjects should be extracted (contrarily to NiftiMapsMasker). Use case:
```
Comment (GaelVaroquaux) [inline nilearn/input_data/multi_nifti_labels_masker.py]:
```suggestion
    different subjects should be extracted (contrarily to NiftiLabelsMasker). Use case:
```
Is that what you meant?

File: nilearn/input_data/multi_nifti_labels_masker.py (hunk 1 above)
Comment (Hororohoruru) [reply nilearn/input_data/multi_nifti_labels_masker.py]: Done. Note that nor the accepted values nor the default are present in the NiftiLabelsMasker docstring.

File: nilearn/input_data/multi_nifti_labels_masker.py (hunk 2 above)
Comment (Hororohoruru) [reply nilearn/input_data/multi_nifti_labels_masker.py]: Done.

File: nilearn/input_data/multi_nifti_labels_masker.py (hunk 3 above)
Comment (Hororohoruru) [reply nilearn/input_data/multi_nifti_labels_masker.py]: Done.

File: nilearn/input_data/multi_nifti_labels_masker.py (hunk 4 above)
Comment (Hororohoruru) [reply nilearn/input_data/multi_nifti_labels_masker.py]: Again, please note that this formatting inconsistencies are also present in the NiftiLabelsMasker object.

Comment (kchawla-pi) [reply nilearn/input_data/multi_nifti_labels_masker.py]: Understood. Changing existing code is a different matter, but for new code, using the `'text {}'.format(str_var)` is the way to go.

File: nilearn/input_data/multi_nifti_labels_masker.py (hunk 10 above)
Comment (kchawla-pi) [reply nilearn/input_data/multi_nifti_labels_masker.py]: "contrary to"

**Comment by Hororohoruru** [comment]
@bthirion Can you have a look? I posted some pointers to work on and the beginning of the PR. Feedback would be great!

File: nilearn/input_data/tests/test_multi_nifti_labels_masker.py (hunk 8 above)
Comment (Hororohoruru) [reply nilearn/input_data/tests/test_multi_nifti_labels_masker.py]: Done!

File: nilearn/input_data/multi_nifti_labels_masker.py [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+        """
+        # We handle the resampling of labels separately because the affine of
+        # the labels image should not impact the extraction of the signal.
+
+        if not hasattr(self, 'mask_img_'):
+            raise ValueError('It seems that {} has not been fitted. '
+                             'You must call fit() before calling transform().'.format
+                             (self.__class__.__name__))
+
+
```
Comment (bthirion) [inline nilearn/input_data/multi_nifti_labels_masker.py:182]: too many lines skipped

File: nilearn/input_data/multi_nifti_labels_masker.py [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+                                       memory_level=self.memory_level,
+                                       verbose=self.verbose)
+
+        if confounds is None:
+            confounds = itertools.repeat(None, len(imgs_list))
+
+        func = self._cache(self.transform_single_imgs,
+                           ignore=['verbose', 'memory', 'memory_level',
+                                   'copy'])
+        data = Parallel(n_jobs=n_jobs)(delayed(func)
```
Comment (bthirion) [inline nilearn/input_data/multi_nifti_labels_masker.py:196]: call it 'region_signals'

File: nilearn/input_data/tests/test_multi_nifti_labels_masker.py [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+    masker11 = MultiNiftiLabelsMasker(labels11_img, resampling_target='data')
+    with pytest.raises(ValueError, match='has not been fitted. '):
+        masker11.transform(fmri11_img)
+
+    # No exception raised here
+    signals11 = masker11.fit().transform(fmri11_img)
+    assert signals11.shape == (length, n_regions)
+
+    # Should work with 4D + 1D input too
+    masker11.transform([fmri01_img, fmri11_img])
```
[the remark this answers is not in the document]
Comment (Hororohoruru) [reply nilearn/input_data/tests/test_multi_nifti_labels_masker.py]: Done!

File: nilearn/input_data/multi_nifti_labels_masker.py [hunk 14]
```diff
@@ -0,0 +1,224 @@
...
+        Region definitions, as one image of labels.
+        background_label: number, optional
+        Label used in labels_img to represent background.
+
+    mask_img: Niimg-like object, optional
+        See http://nilearn.github.io/manipulating_images/input_output.html
+        Mask to apply to regions before extracting signals.
+
+    smoothing_fwhm: float, optional
+        If smoothing_fwhm is not None, it gives the full-width half maximum in
```
[the remark this answers is not in the document]
Comment (Hororohoruru) [reply nilearn/input_data/multi_nifti_labels_masker.py]: Done

File: nilearn/input_data/multi_nifti_labels_masker.py [hunk 15]
```diff
@@ -0,0 +1,224 @@
...
+
+    MultiNiftiLabelsMasker is useful when data from non-overlapping volumes and from
+    different subjects should be extracted (contrary to NiftiLabelsMasker). Use case:
+
+    Parameters
+    ----------
+    labels_img: Niimg-like object
+        See http://nilearn.github.io/manipulating_images/input_output.html
+        Region definitions, as one image of labels.
+        background_label: number, optional
```
[the remark this answers is not in the document]
Comment (Hororohoruru) [reply nilearn/input_data/multi_nifti_labels_masker.py:32]: Oops, fixed

File: nilearn/input_data/multi_nifti_labels_masker.py (hunk 11 above)
Comment (Hororohoruru) [reply nilearn/input_data/multi_nifti_labels_masker.py:182]: Oops 2, fixed

File: nilearn/input_data/multi_nifti_labels_masker.py (hunk 12 above)
Comment (Hororohoruru) [reply nilearn/input_data/multi_nifti_labels_masker.py:196]: Do you mean 'data'?

**Comment by Hororohoruru** [comment]
> Sounds good overall. My biggest concerne is the complexity of all Masker objects.
> It needs to be described in the narrative documentation or in an example or both.
> Also, the docstrings should relate to other maker types to orient the user.

I don't think I get what you are talking about, can you explain more in detail? Also, I think the code from #2247 would be a good starting point for an example.

**Comment by bthirion** [comment]
Explain this new beast here:
https://nilearn.github.io/manipulating_images/masker_objects.html

**Comment by NicolasGensollen** [comment]
Hi @Hororohoruru 
Are you still working on this? AFAICT the last missing piece is a little bit of documentation to explain this new masker. 
Let me know if I can help with that!

**Comment by Hororohoruru** [comment]
Hi @NicolasGensollen !
I am not at the moment, I intended to edit the documentation but I have been busy with other stuff and I don't think I will be coming back to this in the immediate future.

**Comment by NicolasGensollen** [comment]
I started adding a little bit of documentation for `MultiNiftiMasker` and `MultiNiftiLabelsMasker`, but I think that we lack an example in which `MultiNiftiLabelsMasker` could be used. I'm not sure there is already an example that we could improve with this new masker (please correct me if I'm wrong).

A possible solution would be to add a new example from scratch although this conflicts with the objective of reducing the number of examples to decrease doc building time...
Not sure what this example could look like, but here is a small script taken and adapted from #2247 which I used for testing that could provide a starting point...

```python
import time
import numpy as np
from nilearn.input_data import NiftiLabelsMasker, MultiNiftiLabelsMasker
from nilearn import datasets

njobs = 4
development_dataset = datasets.fetch_development_fmri(n_subjects=20)
multiscale = datasets.fetch_atlas_basc_multiscale_2015(version='sym')
atlas_filename = multiscale.scale064

start_multi = time.time()

multi_masker = MultiNiftiLabelsMasker(labels_img=atlas_filename, 
                                      standardize=False,
                                      verbose=0, 
                                      n_jobs=njobs)
time_series_from_multi = multi_masker.fit_transform(development_dataset.func, 
                                                    confounds=development_dataset.confounds)
end_multi = time.time()

single_masker = NiftiLabelsMasker(labels_img=atlas_filename,
                                  standardize=False,
                                  verbose=0)
time_series_from_single = []
for func,conf in zip(development_dataset.func, development_dataset.confounds):
    time_series_from_single.append(single_masker.fit_transform(func, conf))

end_single = time.time()
print("Computation with multi-masker and {} jobs took {}s".format(njobs, end_multi - start_multi))
print("Computation with single-masker took {}s".format(end_single-end_multi))

# Tests
assert len(time_series_from_multi) == 20
assert len(time_series_from_single) == 20
for a,b in zip(time_series_from_multi, time_series_from_single):
    assert a.shape == (168, 64)
    assert b.shape == (168, 64)
    np.testing.assert_array_almost_equal(a,b)
```
```
Computation with multi-masker and 4 jobs took 9.537021398544312s
Computation with single-masker took 23.294074058532715s
```

**Comment by bthirion** [comment]
I guess that MultiNiftiLabelsMasker is mroe useful when we have different atlases for different subjects, am I right ?
(Sorry I never use it).

**Comment by bthirion** [comment]
It would be great to finish it.
You probably want to first re-sync it with the main branch and then focus on adding an example.

**Comment by bthirion** [comment]
And bzw we should also add a MultiNiftiMapsMasker.

**Comment by bthirion** [comment]
The `plot_atlas_comparison` example could benefit from MultiNiftiLabelsMasker if I'm not mistaken.

**Comment by ymzayek** [comment]
Ok working on this
